README.md
# cctest 0.5.2 | [_API Docs_](http://jpcx.github.io/cctest/cctest_8h.html) | [_CHANGELOG_](https://github.com/jpcx/cctest/blob/0.5.2/CHANGELOG.md)

## About

cctest is a tiny test framework for C++17.  

__The API has been simplified since 0.4.2. Please check the usage instructions and CHANGELOG.__

## Usage

Creating a main file (may be seperate or as part of a test source):

```cpp
#define CCTEST_MAIN
#include <cctest.h>
```

Creating test files:

```cpp
#include <cctest.h>
#include <cstring>

TEST("test all the things") {
  static_assert(true, "use static_assert wherever possible!");
  ASSERT(1 == 1); // make a runtime assertion
  auto e = ASSERT_THROWS(std::runtime_error) { // returns thrown error by value
    throw std::runtime_error{"hello"};
  };
  ASSERT(strcmp(e.what(), "hello") == 0);
};

TEST("multiple tests per file are OK") {
  static_assert(true, "runtime assertions are not required");
};
```

_Note: cctest is thread safe, but most tests should be defined at namespace or global scope._

Execution:

``` shell
${CXX} ${CXXFLAGS} -std=c++17 main.cc test0.cc test1.cc -o test
./test

 > this test passes
 > another passing test
 > sample fail test
     FAIL /path/to/src.cc:50
     what: this is an expected failure

 > tests continue after runtime failure

1 Test Failed
```

## Setup

### Standalone Header

Copy `include/cctest.h` into your project and include it.

__note:__  
There should be no compiler warnings for GCC or Clang with `-Wall -Wextra -pedantic`.
Additionally, `#pragma` statements are used to filter Clang's `-Weverything`,
so simple inclusion should be fine in most cases.

### CMake Install

```shell
git clone -b 0.5.2 https://github.com/jpcx/cctest.git
cd cctest

make # mkdir build
     # cmake -Bbuild

# installs header to      ${DESTDIR}${PREFIX}/include/cctest-0.5.2
# installs cmake files to ${DESTDIR}${PREFIX}/lib/cmake/cctest-0.5.2
sudo make install # cmake --install build
```
```cmake
find_package(cctest 0.5 REQUIRED)
target_link_libraries([your target] cctest)
```

### CMake FetchContent

```cmake
include(FetchContent)
FetchContent_Declare(cctest
                     GIT_REPOSITORY https://github.com/jpcx/cctest.git
                     GIT_TAG "0.5.2")
FetchContent_MakeAvailable(cctest)
target_link_libraries([your target] cctest)
```

### CMake add\_subdirectory

```shell
git clone -b 0.5.2 https://github.com/jpcx/cctest.git # or git submodule add
```
```cmake
add_subdirectory([cctest download dir])
target_link_libraries([your target] cctest)
```

## Testing

All API features are self-tested:

```shell
cd [cctest source dir]
make test # mkdir build
          # cmake -Bbuild
          # cmake --build build --target test

```

## Contributing

Contribution is welcome! Please make a pull request.  


CHANGELOG.md
# Changelog

## [Upcoming](https://github.com/jpcx/cctest/compare/0.5.2...devel)

## [0.5.2](https://github.com/jpcx/cctest/releases/tag/0.5.2) \[[diff](https://github.com/jpcx/cctest/compare/0.5.1...0.5.2)\]

### Changed

- Updated copyright year
- Fixed thread safety

## [0.5.1](https://github.com/jpcx/cctest/releases/tag/0.5.1) \[[diff](https://github.com/jpcx/cctest/compare/0.5.0...0.5.1)\]

### Changed

- Updated copyright year
- Improved documentation

### Added

- added built-in `#pragma`-based warning ignores

## [0.5.0](https://github.com/jpcx/cctest/releases/tag/0.5.0) \[[diff](https://github.com/jpcx/cctest/compare/0.4.2...0.5.0)\]

__note:__ The API has been decomplicated and most functionality has been modified.

### Changed

- various semantic and formatting fixes
- `TEST` macro now automatically opens a scope and only accepts one argument
  - example: `TEST("my test") { ASSERT(true); };`
- All test bodies are now void; throw an exception to fail the test.
- thread safety has been added.

### Added

- exceptions thrown within a dynamic predicate are now tracked and printed
- file and line numbers are now tracked when using `TEST`
- added `ASSERT` and `ASSERT_THROWS` macros

### Removed

- removed all static/dynamic test seperation and associated functionality
- removed separate documentation; README is sufficient

## [0.4.2](https://github.com/jpcx/cctest/releases/tag/0.4.2) \[[diff](https://github.com/jpcx/cctest/compare/0.4.1...0.4.2)\]

### Changed

- added `EXCLUDE_FROM_DEFUALT_BUILD` to test target for compatibility
- fixed Changelog typo

## [0.4.1](https://github.com/jpcx/cctest/releases/tag/0.4.1) \[[diff](https://github.com/jpcx/cctest/compare/0.4.0...0.4.1)\]

### Changed

- removed test from default target

## [0.4.0](https://github.com/jpcx/cctest/releases/tag/0.4.0) \[[diff](https://github.com/jpcx/cctest/compare/0.3.0...0.4.0)\]

### Changed

- moved C++17 requirement check from CMake to header

## [0.3.0](https://github.com/jpcx/cctest/releases/tag/0.3.0) \[[diff](https://github.com/jpcx/cctest/compare/0.2.3...0.3.0)\]

### Changed

- made result variables inline, normalized specifier order

## [0.2.3](https://github.com/jpcx/cctest/releases/tag/0.2.3) \[[diff](https://github.com/jpcx/cctest/compare/0.2.2...0.2.3)\]

### Changed

- cleaned up README

## [0.2.2](https://github.com/jpcx/cctest/releases/tag/0.2.2) \[[diff](https://github.com/jpcx/cctest/compare/0.2.1...0.2.2)\]

### Changed

- added `clean` target to `.PHONY` 
- removed patch version from find\_package documentation

## [0.2.1](https://github.com/jpcx/cctest/releases/tag/0.2.1) \[[diff](https://github.com/jpcx/cctest/compare/0.2.0...0.2.1)\]

### Changed

- fixed versioning configure file destination

## [0.2.0](https://github.com/jpcx/cctest/releases/tag/0.2.0) \[[diff](https://github.com/jpcx/cctest/compare/0.1.5...0.2.0)\]

### Changed

- removed extra semicolons from test macros
- moved include directory from root to `include/`
- switched build system to CMake
- updated README.md with new CMake documentation
- added note regarding static tests and macro helpers to README.md
- reverted library entry point back to `cctest.h` (from `cctest/cctest.h`)

### Added

- added `[[noreturn]]` attribute to `dynamic_fail`
- added diff links to CHANGELOG.md
- added root Makefile helper for new CMake build system

## [0.1.5](https://github.com/jpcx/cctest/releases/tag/0.1.5) \[[diff](https://github.com/jpcx/cctest/compare/0.1.4...0.1.5)\]

### Changed

- fixed installation destination documentation

## [0.1.4](https://github.com/jpcx/cctest/releases/tag/0.1.4) \[[diff](https://github.com/jpcx/cctest/compare/0.1.3...0.1.4)\]

### Changed

- changed cctest.h install directory to cctest/cctest.h

## [0.1.3](https://github.com/jpcx/cctest/releases/tag/0.1.3) \[[diff](https://github.com/jpcx/cctest/compare/0.1.2...0.1.3)\]

### Changed

- fixed CHANGELOG.md links; made absolute

## [0.1.2](https://github.com/jpcx/cctest/releases/tag/0.1.2) \[[diff](https://github.com/jpcx/cctest/compare/0.1.1...0.1.2)\]

### Removed

- removed repo link from README.md; localized to webpage

## [0.1.1](https://github.com/jpcx/cctest/releases/tag/0.1.1) \[[diff](https://github.com/jpcx/cctest/compare/0.1.0...0.1.1)\]

### Added

- added link to repo from documentation homepage

### Changed

- fixed private documentation for cctest::detail_::nocvref
- updated two documentation example descriptions
- fixed CHANGELOG.md Upcoming link

## [0.1.0](https://github.com/jpcx/cctest/releases/tag/0.1.0)

### Added

- Basic features, documentation, and tests


test/testcctest.cc
/*                                                                         C++17

                                __                   __
                               /\ \__               /\ \__
                     ___    ___\ \ ,_\    __    ____\ \ ,_\
                    /'___\ /'___\ \ \/  /'__`\ /',__\\ \ \/
                   /\ \__//\ \__/\ \ \_/\  __//\__, `\\ \ \_
                   \ \____\ \____\\ \__\ \____\/\____/ \ \__\
                    \/____/\/____/ \/__/\/____/\/___/   \/__/

                        a tiny test framework for C++17

    Copyright (C) 2020-2023 Justin Collier

      This program is free software: you can redistribute it and/or modify
      it under the terms of the GNU General Public License as published by
      the Free Software Foundation, either version 3 of the License, or
      (at your option) any later version.

      This program is distributed in the hope that it will be useful,
      but WITHOUT ANY WARRANTY; without even the implied warranty of
      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
      GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>. */

#include <chrono>
#include <string>
#include <thread>

#include <cctest.h>

namespace cctest {
namespace detail_ {
extern std::atomic<int> n_tests_failed;
} // namespace detail_
} // namespace cctest

#if defined(__clang__) && __clang__ == 1
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wc++98-compat"
#pragma clang diagnostic ignored "-Wexit-time-destructors"
#pragma clang diagnostic ignored "-Wglobal-constructors"
#endif

namespace testcctest {

/*.----------------------------------------------------------------------------,
 /                             construction tests                            */

static_assert(!std::is_default_constructible_v<cctest::test>);
static_assert(!std::is_copy_constructible_v<cctest::test>);
static_assert(!std::is_copy_assignable_v<cctest::test>);
static_assert(std::is_move_constructible_v<cctest::test>);
static_assert(std::is_move_assignable_v<cctest::test>);

static cctest::test assign = "implicit assignment construction";
static cctest::test movedescr =
    cctest::test{"move-assignment with description"};

static cctest::test op_pass =
    cctest::test{"operator overload", __FILE__, __LINE__} << [] {};
TEST("\"operator overload fail\" fails expectedly") {
  int nbefore = cctest::detail_::n_tests_failed;
  static cctest::test op_fail =
      cctest::test{"operator overload fail", __FILE__, __LINE__}
      << [] { throw std::runtime_error("this is an expected failure"); };
  ASSERT(cctest::detail_::n_tests_failed == nbefore + 1);
};

static cctest::test with_file_and_line_pass =
    cctest::test{"test pass with file and line", __FILE__, __LINE__} << [] {};

TEST("\"test fail with file and line\" fails expectedly") {
  int nbefore = cctest::detail_::n_tests_failed;
  static cctest::test with_file_and_line_fail =
      cctest::test{"test fail with file and line", __FILE__, __LINE__}
      << [] { throw std::runtime_error("this is an expected failure"); };
  ASSERT(cctest::detail_::n_tests_failed == nbefore + 1);
};

TEST("TEST macro usage"){};

TEST("\"TEST macro failure\" fails expectedly") {
  int nbefore = cctest::detail_::n_tests_failed;
  TEST("TEST macro failure") {
    throw std::runtime_error("this is an expected failure");
  };
  ASSERT(cctest::detail_::n_tests_failed == nbefore + 1);
};

TEST("ASSERT passes expectedly") {
  ASSERT(1 == 1);
};

TEST("\"ASSERT fails expectedly\" fails expectedly") {
  int nbefore = cctest::detail_::n_tests_failed;
  TEST("ASSERT fails expectedly") {
    ASSERT(false);
  };
  ASSERT(cctest::detail_::n_tests_failed == nbefore + 1);
};

TEST("ASSERT_THROWS catches and returns an exception") {
  auto e = ASSERT_THROWS(std::runtime_error) {
    throw std::runtime_error{"hey"};
  };
  ASSERT(std::string{e.what()} == "hey");
};

TEST("ASSERT_THROWS throws if no throw") {
  auto e = ASSERT_THROWS(std::runtime_error) {
    ASSERT_THROWS(std::runtime_error){};
  };
  ASSERT(std::string{e.what()} == "[cctest]: throw assertion failed");
};

TEST("ASSERT_THROWS throws if unexpected throw type") {
  auto e = ASSERT_THROWS(std::runtime_error) {
    ASSERT_THROWS(std::runtime_error) {
      throw "bad error";
    };
  };
  ASSERT(std::string{e.what()} == "[cctest]: unexpected throw type");
};

TEST("Tests may be defined in separate threads") {
  static bool t1_tested = false;
  static bool t2_tested = false;

  auto t1 = std::thread{[] {
    TEST("Thread 1") {
      t1_tested = true;
      ASSERT(true);
    };
  }};

  auto t2 = std::thread{[] {
    TEST("Thread 2") {
      t2_tested = true;
      ASSERT(true);
    };
  }};

  t1.join();
  t2.join();
  ASSERT(t1_tested && t2_tested);
};

TEST("expected tests failed") {
  ASSERT(cctest::detail_::n_tests_failed == 4);
  cctest::detail_::n_tests_failed = 0;
};

} // namespace testcctest

#if defined(__clang__) && __clang__ == 1
#pragma clang diagnostic pop
#endif


test/testcctest.main.cc
/*                                                                         C++17

                                __                   __
                               /\ \__               /\ \__
                     ___    ___\ \ ,_\    __    ____\ \ ,_\
                    /'___\ /'___\ \ \/  /'__`\ /',__\\ \ \/
                   /\ \__//\ \__/\ \ \_/\  __//\__, `\\ \ \_
                   \ \____\ \____\\ \__\ \____\/\____/ \ \__\
                    \/____/\/____/ \/__/\/____/\/___/   \/__/

                        a tiny test framework for C++17

    Copyright (C) 2020-2023 Justin Collier

      This program is free software: you can redistribute it and/or modify
      it under the terms of the GNU General Public License as published by
      the Free Software Foundation, either version 3 of the License, or
      (at your option) any later version.

      This program is distributed in the hope that it will be useful,
      but WITHOUT ANY WARRANTY; without even the implied warranty of
      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
      GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>. */

#define CCTEST_MAIN
#include <cctest.h>


testcctest.cc
/////                                                                      c++17
////////////////////////////////////////////////////////////////////////////////
/// @brief cctest - a small static/dynamic testing framework for C++17
//               __                   __
//              /\ \__               /\ \__     auto o = "easy!"
//    ___    ___\ \ ,_\    __    ____\ \ ,_\        << static_pass
//   /'___\ /'___\ \ \/  /'__`\ /',__\\ \ \/        << dynamic_pass
//  /\ \__//\ \__/\ \ \_/\  __//\__, `\\ \ \_       << [] {
//  \ \____\ \____\\ \__\ \____\/\____/ \ \__\           return [...]_pass;
//   \/____/\/____/ \/__/\/____/\/___/   \/__/         };
//
//  cctest C++ Testing
//  Copyright (C) 2020 Justin Collier <m@jpcx.dev>
//
//    This program is free software: you can redistribute it and/or modify
//    it under the terms of the GNU General Public License as published by
//    the Free Software Foundation, either version 3 of the License, or
//    (at your option) any later version.
//
//    This program is distributed in the hope that it will be useful,
//    but WITHOUT ANY WARRANTY; without even the internalied warranty of
//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//    GNU General Public License for more details.
//                                                                            //
//  You should have received a copy of the GNU General Public License        ///
//  along with this program.  If not, see <https://www.gnu.org/licenses/>.  ////
////////////////////////////////////////////////////////////////////////////////

#include <cctest.h>

using namespace cctest;
#define dsc(...) "[cctest] " __VA_ARGS__

namespace test_ {

/*.----------------------------------------------------------------------------,
 /                             construction tests                            */

namespace construction {

test empty{};
test descr{dsc("description-based construction")};
test assign    = dsc("implicit assignment construction");
test copy      = test{};
test copydescr = test{dsc("copy-assignment with description")};

namespace static_ {
test pass{static_pass};
// test fail{static_fail};
test predpass{[]() { return static_pass; }};
test voidpass{[]() {}};
// test predfail{[]() { return static_fail; }};
} // namespace static_

namespace dynamic {
test pass{dynamic_pass};
// test fail{dynamic_fail};
test predpass{[]() { return dynamic_pass; }};
// test predfail{[]() { return dynamic_fail; }};
} // namespace dynamic

} // namespace construction

/*.----------------------------------------------------------------------------,
 /                       member operator overload tests                      */

namespace member_opoverload {

namespace static_ {
test pass = test{dsc("static_pass member operator overload")} << static_pass;
// test fail = test{dsc("static_fail member operator overload")} <<
//             static_fail;
test predpass = test{dsc("static_pass predicate member operator overload")}
                << []() { return static_pass; };
// test predfail = test{dsc("static_fail predicate member operator overload")}
//                 << []() { return static_fail; };
} // namespace static_

namespace dynamic {
test pass = test{dsc("dynamic_pass member operator overload")} << dynamic_pass;
// test fail = test{dsc("dynamic_fail member operator overload")} <<
//             dynamic_fail;
test predpass = test{dsc("dynamic_pass predicate member operator overload")}
                << []() { return dynamic_pass; };
// test predfail =
//     test{dsc("dynamic_fail predicate member operator overload")}
//     << []() { return dynamic_fail; };
} // namespace dynamic

} // namespace member_opoverload

/*.----------------------------------------------------------------------------,
 /                     namespace operator overload tests                     */

namespace namespace_opoverload {

namespace static_ {
test pass = dsc("static_pass namespace operator overload") << static_pass;
// test fail = dsc("static_fail namespace operator overload") <<
//             static_fail;
test predpass = dsc("static_pass predicate namespace operator overload")
                << []() { return static_pass; };
// test predfail = dsc("static_fail predicate namespace operator overload")
//                 << []() { return static_fail; };
} // namespace static_

namespace dynamic {
test pass = dsc("dynamic_pass namespace operator overload") << dynamic_pass;
// test fail = dsc("dynamic_fail namespace operator overload") <<
//             dynamic_fail;
test predpass = dsc("dynamic_pass predicate namespace operator overload")
                << []() { return dynamic_pass; };
// test predfail =
//     dsc("dynamic_fail predicate namespace operator overload")
//     << []() { return dynamic_fail; };
} // namespace dynamic

} // namespace namespace_opoverload

/*.----------------------------------------------------------------------------,
 /                         test resulter combinations                        */

namespace combinations {

test comb0 = test{dsc("test -> static_pass -> static_pass_pred -> dynamic_pass "
                      "-> dynamic_pass_pred")}
             << static_pass << []() { return static_pass; } << dynamic_pass
             << []() { return dynamic_pass; };
test comb1 =
    dsc("(const char * << static_pass) -> static_pass_pred -> dynamic_pass "
        "-> dynamic_pass_pred")
    << static_pass << []() { return static_pass; } << dynamic_pass
    << []() { return dynamic_pass; };

test comb2 = test{dsc("test -> dynamic_pass -> static_pass")} << dynamic_pass
                                                              << static_pass;

test comb3 = test{dsc("dynamic_pass -> static_pass")} << dynamic_pass
                                                      << static_pass;

} // namespace combinations

} // namespace test_

/*.----------------------------------------------------------------------------,
 /                             test helper macros                            */

TEST(cctest.macros, "provides testing helper macros")
    << STATIC_REQUIRE(true) << DYNAMIC_REQUIRE(true);

TEST(cctest.macros, "facilitates static scopes") << [] { STATIC_CHECK(true); };

TEST(cctest.macros, "facilitates dynamic scopes") << [] {
  DYNAMIC_CHECK(true);
  CLOSE_SCOPE;
};

TEST(cctest.macros, "facilitates static + dynamic scopes") << [] {
  STATIC_CHECK(true);
  DYNAMIC_CHECK(true);
  CLOSE_SCOPE;
};

int
main(int, char **) {
  cctest::success();
}


include/cctest.h
#ifndef CCTEST_H_INCLUDED
#define CCTEST_H_INCLUDED
/*                                                                         C++17

                                __                   __
                               /\ \__               /\ \__
                     ___    ___\ \ ,_\    __    ____\ \ ,_\
                    /'___\ /'___\ \ \/  /'__`\ /',__\\ \ \/
                   /\ \__//\ \__/\ \ \_/\  __//\__, `\\ \ \_
                   \ \____\ \____\\ \__\ \____\/\____/ \ \__\
                    \/____/\/____/ \/__/\/____/\/___/   \/__/

                        a tiny test framework for C++17

    Copyright (C) 2020-2023 Justin Collier

      This program is free software: you can redistribute it and/or modify
      it under the terms of the GNU General Public License as published by
      the Free Software Foundation, either version 3 of the License, or
      (at your option) any later version.

      This program is distributed in the hope that it will be useful,
      but WITHOUT ANY WARRANTY; without even the implied warranty of
      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
      GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>. */

#if (!defined(__cplusplus) || __cplusplus < 201703L)
#error "cctest requires at least C++17"
#endif

/// \file
/// \brief cctest is a tiny test framework for C++17 [single-header]
///
/// Creating test files:
///
/// \code
///   #include <cctest.h>
///   #include <cstring>
///
///   TEST("test all the things") {
///     static_assert(true, "use static_assert wherever possible!");
///     ASSERT(1 == 1); // make a runtime assertion
///     auto e = ASSERT_THROWS(std::runtime_error) { // returns error by value
///       throw std::runtime_error{"hello"};
///     };
///     ASSERT(strcmp(e.what(), "hello") == 0);
///   };
///
///   TEST("multiple tests per file are OK") {
///     static_assert(true, "runtime assertions are not required");
///   };
/// \endcode
///
/// Creating a main file:
///
/// \code
///   #define CCTEST_MAIN
///   #include <cctest.h>
/// \endcode
///
/// Execution:
///
/// \code
///   ${CXX} ${CXXFLAGS} -std=c++17 main.cc test0.cc test1.cc -o test
///   ./test
/// \endcode

#include <atomic>
#include <iostream>
#include <mutex>
#include <string_view>

#if defined(__clang__) && __clang__ == 1
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wc++98-compat"
#pragma clang diagnostic ignored "-Wc++98-compat-pedantic"
#pragma clang diagnostic ignored "-Wpadded"
#endif

namespace cctest {

/*.----------------------------------------------------------------------------,
 /                             internal details                              */

namespace detail_ {
extern std::atomic<int> n_tests_failed;

namespace is_void_fn_ {
/// sfinae fallback false
template <class, class = void>
struct sfinae {
  static constexpr bool result = false;
};
/// sfinae valid if F returns void
template <class F>
struct sfinae<
    F, std::enable_if_t<std::is_same_v<decltype(std::declval<F>()()), void>>> {
  static constexpr bool result = true;
};
} // namespace is_void_fn_
template <class F>
inline constexpr bool is_void_fn = is_void_fn_::sfinae<F>::result;
template <class Exception>
struct assert_throws {
  template <class Op, std::enable_if_t<is_void_fn<Op>>* = nullptr>
  inline Exception
  operator<<(Op&& op) const {
    try {
      op();
    } catch (Exception const& except) { return except; } catch (...) {
      throw std::runtime_error("[cctest]: unexpected throw type");
    }
    throw std::runtime_error("[cctest]: throw assertion failed");
  }
};
} // namespace detail_

/*.----------------------------------------------------------------------------,
 /                               cctest::test                                */

/// assertion class and description logger
struct test {
  constexpr test()            = delete;
  constexpr test(test const&) = delete;
  constexpr test&
  operator=(test const&) = delete;
  test(test&&)           = default;
  test&
  operator=(test&&) = default;
  /// default destructor
  ~test() noexcept = default;
  /// load the description
  test(char const* description) : description_(description) {
  }
  /// load the description, filename, and line number
  test(char const* description, char const* file, int line)
      : description_(description), file_(file), line_(line) {
  }

  /// invoke and assert a test body
  template <class Fn, std::enable_if_t<detail_::is_void_fn<Fn>>* = nullptr>
  inline test&&
  operator<<(Fn&& pred) noexcept {
    try {
      pred();
      pass();
    } catch (...) { fail(std::current_exception()); }
    return std::move(*this);
  }

 private:
  std::string description_        = "";
  std::string file_               = "";
  int line_                       = -1;
  inline static std::mutex m_out_ = {};
  void
  pass() const noexcept {
    std::scoped_lock<std::mutex> _{m_out_};
    std::cout << get_description();
  }
  /// logs a failure message
  void
  fail(std::exception_ptr exception = nullptr) const noexcept {
    ++detail_::n_tests_failed;
    auto descr = get_description();
    descr += "\n     \033[01m\033[31mFAIL\033[0m\033[0m ";
    descr += file_;
    descr += ":";
    descr += std::to_string(line_);
    descr += '\n';
    try {
      if (exception != nullptr)
        std::rethrow_exception(exception);
      else
        throw std::runtime_error{""};
    } catch (std::exception const& e) {
      descr += "     what: ";
      descr += e.what();
      descr += "\n";
    } catch (...) {}
    {
      std::scoped_lock<std::mutex> _{m_out_};
      std::cout << descr;
    }
  }
  /// creates an initial description
  inline std::string
  get_description() const noexcept {
    return std::string{"\n\033[90m>\033[0m "} + description_;
  }
};

/*.----------------------------------------------------------------------------,
 /                   main execution success message logger                   */

/// prints "All Tests Passed!" in bold green to the console to indicate success
inline void
success() noexcept {
  std::cout << "\n\n\033[01m\033[32mAll Tests Passed!\033[0m\033[0m\n";
}

/// prints "[n] Tests Failed" in bold red to the console to indicate success
inline void
failure() noexcept {
  std::cout << "\n\n\033[01m\033[31m" << detail_::n_tests_failed << " Test"
            << (detail_::n_tests_failed > 1 ? "s" : "")
            << " Failed\033[0m\033[0m\n";
}

} // namespace cctest

/*.----------------------------------------------------------------------------,
 /                                   main                                    */

#ifndef CCTEST_DISABLE_MACROS
#ifdef TEST
#error \
    "TEST macro is already in use. please remove the library responsible or " \
    "#define CCTEST_DISABLE_MACROS"
#endif
#ifdef ASSERT
#error \
    "ASSERT macro is already in use. please remove the library responsible or " \
    "#define CCTEST_DISABLE_MACROS"
#endif
#ifdef ASSERT_THROWS
#error "ASSERT_THROWS macro is already in use. please remove the library " \
       "responsible or #define CCTEST_DISABLE_MACROS"
#endif

#define CCTEST_IMPL_CAT(_a, _b)   CCTEST_IMPL_CAT_X(_a, _b)
#define CCTEST_IMPL_CAT_X(_a, _b) _a##_b

#define TEST(_descr)                                        \
  static cctest::test CCTEST_IMPL_CAT(test_at_, __LINE__) = \
      cctest::test{_descr, __FILE__, __LINE__} << []

#define ASSERT(...)   \
  if (!(__VA_ARGS__)) \
    throw std::runtime_error { "assertion failed: " #__VA_ARGS__ }
#define ASSERT_THROWS(...) cctest::detail_::assert_throws<__VA_ARGS__>{} << [&]

#ifdef CCTEST_MAIN
namespace cctest {
namespace detail_ {
std::atomic<int> n_tests_failed = 0;
} // namespace detail_
} // namespace cctest

int
main(int, char**) {
  if (cctest::detail_::n_tests_failed == 0) {
    cctest::success();
    return 0;
  } else {
    cctest::failure();
    return 1;
  }
}
#endif
#endif

#if defined(__clang__) && __clang__ == 1
#pragma clang diagnostic pop
#endif

#endif
